File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.cpp
#include <string>
#include <vector>
#include <queue>
#define X first
#define Y second

using namespace std;

vector<vector<int>> table2, game_board2;
vector<vector<vector<int>>> puzzle_p, board_p;
vector<bool> visit;
int dx[4] = {-1,0,1,0}, dy[4] = {0,-1,0,1};
int n;

void split_pieces(int a, int b, vector<vector<int>> &t, 
                  vector<vector<int>> &t2, 
                  bool is_pz){
    queue<pair<int,int>> q;
    q.push({a,b}); t2[a][b] = 0;
    
    int sr = a, er = a, sc = b, ec = b;
    
    while(!q.empty()){
        auto cur = q.front(); q.pop();
        int x = cur.X, y = cur.Y;
        
        for(int i = 0; i < 4; i++){
            int nx = x + dx[i], ny = y + dy[i];
            if(nx < 0 || ny < 0 || nx >= n || ny >= n) continue;
            if(t2[nx][ny] != is_pz) continue;
            t2[nx][ny] = !is_pz; q.push({nx,ny});
            
            if(nx < sr) sr = nx;
            if(ny < sc) sc = ny;
            if(nx > er) er = nx;
            if(ny > ec) ec = ny;
        }
    }
    
    vector<vector<int>> v;
    
    for(int i = sr; i <= er; i++){
        vector<int> r;
        for(int j = sc; j <= ec; j++) {
            if(is_pz) r.push_back(t[i][j]);
            else r.push_back(!t[i][j]);
        }
        v.push_back(r);
    }
    if(is_pz) puzzle_p.push_back(v);
    else board_p.push_back(v);
}

vector<vector<int>> rotate_p(vector<vector<int>> cur){
    int r = cur.size(), c = cur[0].size();
    vector<vector<int>> ret(c, vector<int>(r, 0));
    for(int i = 0; i < r; i++){
        for(int j = 0; j < c; j++){
            ret[j][i] = cur[r - i - 1][j];
        }
    }
    return ret;
}

int comp_cnt(int &br, int &bc, vector<vector<int>> &bp, vector<vector<int>> &pp){
    int cnt = 0;
    for(int i = 0; i < br; i++){
        for(int j = 0; j < bc; j++) {
            if(bp[i][j] != pp[i][j]) return 0;
            if(bp[i][j] == 1) cnt++;
        }   
    }
    return cnt;
}

int compare_p(vector<vector<int>> &bp, vector<vector<int>> pp, int idx){
    int br = bp.size(), bc = bp[0].size();
    int pr = pp.size(), pc = pp[0].size(), cnt = 0;
    
    if(br == pr && bc == pc){
        if(br == bc) { // 정사각형 -> 0 ~ 3 회전 비교
            for(int i = 0; i < 4; i++) {
                if(i > 0) pp = rotate_p(pp);
                cnt = comp_cnt(br, bc, bp, pp);
                if(cnt) {
                    visit[idx] = 1;
                    return cnt;
                }
            }
        } else {
            cnt = comp_cnt(br, bc, bp, pp); // 0 회전
            if(cnt) {
                visit[idx] = 1;
                return cnt;
            }
            pp = rotate_p(rotate_p(pp)); // 2 회전
            cnt = comp_cnt(br, bc, bp, pp);
        }
    } else {
        if(br != pc || bc != pr) return 0; // 1 회전 해도 안됨 -> 불가능
        pp = rotate_p(pp); // 1 회전
        cnt = comp_cnt(br, bc, bp, pp);
        if(cnt) {
            visit[idx] = 1;
            return cnt;
        }
        pp = rotate_p(rotate_p(pp)); // 3 회전
        cnt = comp_cnt(br, bc, bp, pp);
    }
    
    if(cnt) visit[idx] = 1;
    return cnt;
}

int solution(vector<vector<int>> game_board, vector<vector<int>> table) {
    n = table.size();
    game_board2 = game_board;
    table2 = table;
    int answer = 0;
    
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            if(table2[i][j]) split_pieces(i, j, table, table2, 1);
            if(!game_board2[i][j]) split_pieces(i, j, game_board, game_board2, 0);
        } 
    }    
    
    visit.resize(puzzle_p.size(), 0);
    int idx = 0;
    
    for(auto bp : board_p){
        idx = 0;
        for(auto pp : puzzle_p){
            if(visit[idx++]) continue;
            int tmp = compare_p(bp, pp, idx-1);
            answer += tmp;
            if(tmp) break;
        }
    }
    
    
    return answer;
}

File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.java
import java.util.*;


class Solution {
    boolean[][] visit;
    int N;
    int[] dx = {-1,0,1,0};
    int[] dy = {0,-1,0,1};
    List<Piece> boardPieces = new ArrayList<>();
    List<Piece> puzzlePieces = new ArrayList<>();
    
    public int solution(int[][] game_board, int[][] table) {
        int answer = 0;
        
        N = game_board.length;
        visit = new boolean[N][N];
        
        // 조각 분리
        for(int i = 0; i < N; i++){
            for(int j = 0; j < N; j++){
                if(game_board[i][j] == 0 && !visit[i][j]) 
                    splitPieces(new int[] {i, j}, 0, visit, game_board, boardPieces);
            }
        }
        
        visit = new boolean[N][N];
        
        for(int i = 0; i < N; i++){
            for(int j = 0; j < N; j++){
                if(table[i][j] == 1 && !visit[i][j]) 
                    splitPieces(new int[] {i, j}, 1, visit, table, puzzlePieces);
            }
        }
        
        int idx = 0;
        boolean[] used = new boolean[puzzlePieces.size()];
        
        for(Piece b : boardPieces){
            idx = 0;
            for(Piece p : puzzlePieces){
                if(used[idx++]) continue;
                int cnt = comparePieces(b, p);
                if(cnt == 0) continue;
                System.out.println("b : \n" + b);
                System.out.println("p : \n" + p);
                System.out.println("count : " + cnt);
                
                answer += cnt;
                used[idx - 1] = true;
                break;
            }
        }        
        
        return answer;
    }
    
    public int comparePieces(Piece b, Piece p){
        for(int i = 0; i < 4; i++){
            int br = b.arr.length, bc = b.arr[0].length;
            int pr = p.arr.length, pc = p.arr[0].length;
            if(br != pr && br != pc && bc != pc && bc != pr) return 0;
            if(br == pr && bc == pc) {
                int cnt = countCells(b, p);
                if(cnt > 0) return cnt;
            }
            p.rotate();
        }
        return 0;
    }
    
    public int countCells(Piece b, Piece p){
        int r = b.arr.length, c = b.arr[0].length;
        int cnt = 0;
        
        for(int i = 0; i < r; i++){
            for(int j = 0; j < c; j++){
                if(b.arr[i][j] == p.arr[i][j]) 
                    cnt += p.arr[i][j];
                else return 0;
            }
        }
        
        return cnt;
    }
    
    public void splitPieces(int[] pos, int target, boolean[][] visit, int[][] board, List<Piece> pieces){
        Queue<int[]> q = new LinkedList<>();
        visit[pos[0]][pos[1]] = true;
        q.add(pos);
        
        int rowMin = pos[0], rowMax = pos[0], colMin = pos[1], colMax = pos[1]; 
        
        while(!q.isEmpty()){
            int[] cur = q.poll();
            for(int d = 0; d < 4; d++){
                int nx = cur[0] + dx[d], ny = cur[1] + dy[d];
                if(nx < 0 || ny < 0 || nx >= N || ny >= N) continue;
                if(visit[nx][ny] || board[nx][ny] != target) continue;
                q.add(new int[] {nx, ny});
                visit[nx][ny] = true;
                
                if(nx < rowMin) rowMin = nx;
                if(nx > rowMax) rowMax = nx;
                if(ny < colMin) colMin = ny;
                if(ny > colMax) colMax = ny;
            }
        }
      
        int r = rowMax - rowMin + 1, c = colMax - colMin + 1;
        int[][] tmp = new int[r][c];
        
        for(int i = 0; i < r; i++)
           for(int j = 0; j < c; j++)
                if(board[rowMin + i][colMin + j] == target) tmp[i][j] = 1;
        
        pieces.add(new Piece(tmp));
    }
}



class Piece {
    public int[][] arr;
    public Piece(int[][] arr){
        int r = arr.length, c = arr[0].length; 
        this.arr = new int[r][c];

        for(int i = 0; i < r; i++)
            this.arr[i] = arr[i].clone();
    }
    
    public void rotate(){
        int r = this.arr.length, c = this.arr[0].length;
        int[][] tmp = new int[c][r];
        
        for(int i = 0; i < r; i++){
            for(int j =0; j < c; j++) {
                tmp[j][i] = arr[r-i-1][j];
            }
        }
        
        this.arr = tmp;
    }
    
    public String toString(){
        StringBuilder sb = new StringBuilder();
        sb.append("\n");
        
        for(int i =0 ; i< arr.length;i ++){
            for(int j =0;j < arr[0].length;j ++){
                sb.append(arr[i][j]);
            }
            sb.append("\n");
        }
        
        sb.append("\n");
        
        return sb.toString();
    }
}


